File: maximum_subsequence_score.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @param {number} k
 * @return {number}
 */
var maxScore = function(nums1, nums2, k) {
    const pairs = nums1.map((x,i) => [x, nums2[i]]).sort((a,b) => b[1] - a[1]);
    let total = 0, res = 0;
    const queue = new MinPriorityQueue();
    
    for (const [n1, n2] of pairs) {
        queue.enqueue(n1);
        total += n1;

        if (queue.size() === k) {
            res = Math.max(res, total * n2);
            total -= queue.dequeue().element;
        }
    }

    return res;
};

File: maximum_subsequence_score.py
class Solution:
    def maxScore(self, nums1: List[int], nums2: List[int], k: int) -> int:
        pairs = sorted(zip(nums1, nums2), key=lambda p: -p[1])

        total = res = 0
        heap = []

        for pair in pairs:
            nums1, nums2 = pair
            heapq.heappush(heap, nums1)
            total += nums1

            if len(heap) > k:
                total -= heapq.heappop(heap)
            
            if len(heap) == k:
                res = max(res, total * nums2)
        
        return res
